File: solutions/python/list-ops/1/list_ops.py
def append(list1, list2):
    list1.extend(list2)
    return list1


def concat(lists):
    return [item for sublist in lists for item in sublist]


def filter(function, list):
    return [ item for item in list if function(item)]


def length(list):
    return len(list)


def map(function, list):
    return [ function(item) for item in list ]


def foldl(function, list, initial):
    out = initial
    for item in list:
        out = function(out, item)
    return out


def foldr(function, list, initial):
    out = initial
    for item in reversed(list):
        out = function(out, item)
    return out


def reverse(list):
    return list[::-1]

File: solutions/java/list-ops/1/src/main/java/ListOps.java
import java.util.ArrayList;
import java.util.List;
import java.util.function.BiFunction;
import java.util.function.Function;
import java.util.function.Predicate;

class ListOps {

    static <T> List<T> append(List<T> list1, List<T> list2) {
        List<T> result = new ArrayList<T>(list1.size() + list2.size());
        result.addAll(list1);
        result.addAll(list2);
        return result;
    }

    static <T> List<T> concat(List<List<T>> listOfLists) {
        return listOfLists.stream()
            .flatMap(List::stream)
            .collect(ArrayList::new, ArrayList::add, ArrayList::addAll);
    }

    static <T> List<T> filter(List<T> list, Predicate<T> predicate) {
        return list.stream().filter(predicate).toList();
    }

    static <T> int size(List<T> list) {
        return list.size();
    }

    static <T, U> List<U> map(List<T> list, Function<T, U> transform) {
        return list.stream().map(transform).toList();
    }

    static <T> List<T> reverse(List<T> list) {
        return list.stream().collect(
            ArrayList::new, 
            (acc, item) -> acc.add(0, item), 
            (acc1, acc2) -> acc1.addAll(0, acc2));
    }

    static <T, U> U foldLeft(List<T> list, U initial, BiFunction<U, T, U> f) {
        U current = initial;
        for (T item: list) {
            current = f.apply(current, item);
        }
        return current;
    }

    static <T, U> U foldRight(List<T> list, U initial, BiFunction<T, U, U> f) {
        U current = initial;
        for (T item: ListOps.reverse(list)) {
            current = f.apply(item, current);
        }
        return current;
    }

    private ListOps() {
        // No instances.
    }

}
